cogito/ghcommitsink.go

```diff
@@ -93,9 +93,21 @@ func ghMakeContext(request PutRequest) string {
 }
 
 func apiEndpoint(h string) string {
-	hostname := strings.ToLower(h)
+	hostname := normalizeHostname(h) // not really required; already called during protocol init
+	if hostname == github.GhDefaultHostname {
+		return github.API
+	}
 	if strings.Contains(hostname, "127.0.0.1") {
 		return fmt.Sprintf("http://%s", hostname)
 	}
-	return github.API
+	return fmt.Sprintf("https://%s/api/v3", hostname)
+}
+
+func normalizeHostname(h string) string {
```

Comment:
of course, I will rewrite this in a cleaner way; probably using TrimFunction...

Reply:
There must exist a go lib to just parse the URL and extract the hostname.

Reply:
Hmm... yes it exists but I don't see how I can use it here. In `source.github_hostname` we expect the hostname it self, not URL e.g a string like `github.com`, `my-ghe.com` or `127.0.1:5435`. I want to add this normalizedHostname function (inspired by github cli tool code) to handle the cases when a user writes configures this variable in a way we didn't expect i.e by adding `/` or `/api/v1` to it...

Why return an error in that case to the user if we can easily recover from it?

Reply:
You don't have to return an error because you use a tool that can return an error. Just handle it. :-)
Pseudo code (based on how python urlparse handle `"foo.bar.baz"`):
```
  parsed_url, err := url.parse(h)
  if err != nil {
    return h
  }
  if parsed_url.hostname != "" {
    return parse_url.hostname
  }
  if parsed_url.path != "" {
    return parsed_url.path.split("/")[0]
  }
```

Reply:
My understanding is that you would need to inspect an error; so already your first check would be wrong.

We expect the user to configure: `github_hostname` by configuring `my-ghe.com` or test `127.0.1:1234` the url.parse would just return an error for both because there is no protocol configured. So in the first if statement I would need to cover inspect the error and continue if protocol is missing; and we don't even want the user to configure the protocol.

Note, if I expect the user to configure the protocol also so some kind of: `github_base_url` then the other part of the code needs to somehow strip the protocol. Additionally having `github_base_url` is useless because there is no easy way to construct the API endpoint for all 3 cases (default, test and enterprise).

Reply:
All this is a bit confusing and I agree with Olivier that _if_ we could use the Go stdlib function URL parse to detect errors it would be better.

To help understanding the intent, we can start from documenting the function. What is the intent of the function? Can we put in the documentation two examples input/output to help the reader?

Reply:
So in this iteration I removed this normalization function with two goals:
1. first to show you that its not really required by anyone. Documentation should be clear enough:
```
  GitHub hostname. This allows to post commit statuses to repositories hosted by GitHub Enterprise (GHE) instances. For example: github.mycompany.org will be expanded by cogito to https://github.mycompany.org/api/v3.
```
if you configure the `source.github_hostname` correctly (no protocol, no API prefix, no trailing slash) it will just work.
2. after the point 1 is clear. We can now discuss bringing the normalization function back in the PR if we agree that it brings value and I will try to write it in a better way.

My point of view was: lets help the user in case it misconfigures the cogito by not failing if the most common mistakes are done:
- protocol added to the hostname: `source.github_hostname = https://github.mycompany.org`
- the trailing slash: `source.github_hostname = https://github.mycompany.org/api/v3/`
- API prefix already added: `source.github_hostname = github.mycompany.org/api/v3/`
- or both protocol and API prefix: `source.github_hostname = https://github.mycompany.org/api/v3/`

If you see benefit in the this normalization function, I will try to bring it back.

Opinions?